### PR TITLE
Comments: support live comments in dashboards and annotations

### PR DESCRIPTION
**Pull request by FZambia**
This pr adds 2 kinds of live comments.

1. Dashboard-wide comments, toggle `dashboardComments`
2. Annotation comments, toggle `annotationComments`

Questions:

* [x] ~~Naming - currently we call sth here `chats`, sth `discussions`, need to stick with one.~~ Stick with `comments`
* [x] CSS styles for components - some of them inlined, some put to the wrong file maybe
* [x] Check runtime/data packages usage
* [ ] Discuss current permission checks
* [x] Placement of dashboard comment icon

TODO (future?):

* [ ]  Re-initialize widget on connection re-establishement.
* [ ]  Message pagination, at the moment we only display 100 last messages
* [ ]  Avoid scrolling upon receiving new message from Live if scroll is not on the bottom (otherwise new message will result to undesired jumps towards the input) 
* [ ] Sent comments not displayed with Grafana Live disabled - need to append comment from request response to gracefully handle disabled Live
* [ ] In case of missing edit permission hide the comment input
* [ ] Delete comments if an object id they are binded to deleted?
* [ ]  We can clean up old history automatically, is it fine to not clean up at all by default from a legal point of view as we don't have delete API yet.

Demo:

https://user-images.githubusercontent.com/1196565/153209001-cbc09997-8b32-4ec1-9fd7-89f38b889cc8.mp4

### CODE REVIEW

File: public/app/features/chat/ChatIndex.tsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { Chat } from '../../../../packages/grafana-runtime';
+
+export default function ChatIndex() {
+  const navModel = useNavModel('chat');
+
+  return (
+    <Page navModel={navModel}>
+      <Page.Contents className="chat-wrapper">
+        {/* TODO: pass proper organization ID as objectId */}
+        <Chat contentTypeId={1} objectId={'1'} />
```
Comment (FZambia) [inline public/app/features/chat/ChatIndex.tsx:13]: Can we somehow access orgId here?

File: packages/grafana-ui/src/themes/GlobalStyles/page.ts [hunk 2]
```diff
@@ -117,5 +117,79 @@ export function getPageStyles(theme: GrafanaTheme2) {
       margin-left: ${theme.spacing(1)};
       margin-top: ${theme.spacing(0.5)};
     }
+
+    .chat-wrapper {
```
Comment (ryantxu) [inline packages/grafana-ui/src/themes/GlobalStyles/page.ts:121]: :thinking: in general -- we try to avoid adding raw css to global styles.  Ideally this can be moved to the component that will use the styles

File: pkg/api/index.go [hunk 3]
```diff
@@ -198,6 +198,16 @@ func (hs *HTTPServer) getNavTree(c *models.ReqContext, hasEditPerm bool) ([]*dto
 		Children:   dashboardChildLinks,
 	})
 
+	navTree = append(navTree, &dtos.NavLink{
```
Comment (ryantxu) [inline pkg/api/index.go:201]: feature flag?

**Comment by torkelo** [comment]
Love dig into this more later in the week, really think there is something very useful here. 

But I would not frame this as chat, think chat should be best handled by chat tools like slack etc. I think of this as something very similar/identical to Google doc comments makes a lot of sense for dashboards, alert rules, reports, incidents, etc.

File: packages/grafana-ui/src/components/ContextMenu/ContextMenu.tsx [hunk 4]
```diff
@@ -11,14 +11,16 @@ export interface ContextMenuProps {
   y: number;
   /** Callback for closing the menu */
   onClose?: () => void;
+  /** Enable focus on first Menu element on openning the menu */
+  focusOnOpen?: boolean;
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/ContextMenu/ContextMenu.tsx:15]: The changes for ContextMenu should be in their own PR -- so the simple behavior change is clearly noted in the commit history.

File: packages/grafana-data/src/text/markdown.ts [hunk 5]
```diff
@@ -42,3 +42,25 @@ export function renderTextPanelMarkdown(str?: string, options?: RenderMarkdownOp
 
   return sanitizeTextPanelContent(html);
 }
+
+// Markdown for Grafana Chats only!
+export function renderChatMarkdown(str?: string, options?: RenderMarkdownOptions): string {
```
Comment (ryantxu) [inline packages/grafana-data/src/text/markdown.ts:47]: As a simple first step... move this function into:

packages/grafana-runtime/src/components/Chat.tsx

File: packages/grafana-runtime/src/components/Chat.tsx [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+  event: string;
+  messageCreated: Message;
+}
+
+export interface ChatState {
+  messages: Message[];
+  value: string;
+}
+
+export class Chat extends PureComponent<ChatProps, ChatState> {
```
Comment (ryantxu) [inline packages/grafana-runtime/src/components/Chat.tsx:40]: do we need this accessible to plugins (directly) yet?  If not, we should move all frontend changes to:

https://github.com/grafana/grafana/tree/main/public/app/features/chat

and then add @grafana/grafana-edge-squad to CODEOWNERS for that folder.

File: public/app/features/chat/ChatIndex.tsx (hunk 1 above)
Comment (ryantxu) [reply public/app/features/chat/ChatIndex.tsx:13]:
```
import { contextSrv } from 'app/core/services/context_srv';

....

contextSrv.user.orgId


```

File: public/app/features/chat/ChatIndex.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import Page from 'app/core/components/Page/Page';
+import { useNavModel } from 'app/core/hooks/useNavModel';
+import { Chat } from '../../../../packages/grafana-runtime';
```
Comment (ryantxu) [inline public/app/features/chat/ChatIndex.tsx:4]: simplifed after everything is moved to this folder

File: public/app/features/chat/routes.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { SafeDynamicImport } from 'app/core/components/DynamicImports/SafeDynamicImport';
+import { config } from 'app/core/config';
+import { RouteDescriptor } from 'app/core/navigation/types';
+
+const chatRoutes = [
```
Comment (ryantxu) [inline public/app/features/chat/routes.tsx:5]: I don't think we need to start with the org wide chat page.  That feels much less useful than a commenting system within panels/dashbaords

File: pkg/api/index.go (hunk 3 above)
Comment (FZambia) [reply pkg/api/index.go]: done, now using feature flag that enables discussions in general (migrations), and separate flags for each org, dashboard, annotation discussions.

File: packages/grafana-runtime/src/components/Chat.tsx (hunk 6 above)
Comment (FZambia) [reply packages/grafana-runtime/src/components/Chat.tsx]: done

File: public/app/features/chat/ChatIndex.tsx (hunk 7 above)
Comment (FZambia) [reply public/app/features/chat/ChatIndex.tsx]: done

File: public/app/features/chat/routes.tsx (hunk 8 above)
Comment (FZambia) [reply public/app/features/chat/routes.tsx:5]: Uses a separate feature toggle now

File: packages/grafana-data/src/text/markdown.ts (hunk 5 above)
Comment (FZambia) [reply packages/grafana-data/src/text/markdown.ts]: Done in 52906830fb1ae0dff4049d7053c255f3aa59a6cd

So I placed func near Chat component, this required adding `marked` to Grafana package.json. Also, using second argument in `marked` function to set custom options per-call and avoid setting options on global marked instance. Hope this is fine.

Comment (FZambia) [reply packages/grafana-data/src/text/markdown.ts]: CI now fails – and I don't understand why :( https://github.com/grafana/grafana/runs/5107487642?check_suite_focus=true

File: public/app/features/chat/Chat.tsx [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+  messageCreated: Message;
+}
+
+export interface ChatState {
+  messages: Message[];
+  value: string;
+}
+
+function renderChatMarkdown(str?: string): string {
+  const html = marked(str || '', {
```
Comment (ryantxu) [inline public/app/features/chat/Chat.tsx:42]: rather than use marked directly, should should likey just use:
```
import { renderTextPanelMarkdown } from '@grafana/data';
```
and respect the sanatize flag from `config.disableSanitizeHtml`  see:

https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/text/TextPanel.tsx#L47

File: package.json [hunk 10]
```diff
@@ -306,6 +307,7 @@
     "lodash": "4.17.21",
     "logfmt": "^1.3.2",
     "lru-cache": "6.0.0",
+    "marked": "4.0.10",
```
Comment (ryantxu) [inline package.json:310]: not needed if we import `renderTextPanelMarkdown`

File: packages/grafana-data/src/text/markdown.ts [hunk 11]
```diff
@@ -42,3 +42,15 @@ export function renderTextPanelMarkdown(str?: string, options?: RenderMarkdownOp
 
   return sanitizeTextPanelContent(html);
 }
+
+export function renderChatMarkdown(str?: string): string {
```
Comment (ryantxu) [inline packages/grafana-data/src/text/markdown.ts:46]: necessary?   do we want/need something different than what we are doing in the text panel?

File: pkg/api/api.go [hunk 12]
```diff
@@ -80,6 +80,8 @@ func (hs *HTTPServer) registerRoutes() {
 	r.Get("/live/pipeline", reqGrafanaAdmin, hs.Index)
 	r.Get("/live/cloud", reqGrafanaAdmin, hs.Index)
 
+	r.Get("/chat", reqSignedIn, hs.Index)
```
Comment (ryantxu) [inline pkg/api/api.go:83]: lets remove this to start -- org wide chat seems too overwhelming!

File: pkg/services/featuremgmt/registry.go [hunk 13]
```diff
@@ -126,5 +126,25 @@ var (
 			Description: "Highlight Enterprise features",
 			State:       FeatureStateStable,
 		},
+		{
+			Name:        "liveDiscussions",
+			Description: "Enable Live discussions feature",
+			State:       FeatureStateAlpha,
+		},
+		{
+			Name:        "liveOrgDiscussions",
```
Comment (ryantxu) [inline pkg/services/featuremgmt/registry.go:135]: I am skeptical of this flavor even behind a feature flag

File: pkg/services/featuremgmt/registry.go [hunk 14]
```diff
@@ -126,5 +126,25 @@ var (
 			Description: "Highlight Enterprise features",
 			State:       FeatureStateStable,
 		},
+		{
+			Name:        "liveDiscussions",
```
Comment (ryantxu) [inline pkg/services/featuremgmt/registry.go:130]: Is this one necessary?  liveDashboardDiscussions | liveAnnotationDiscussions?

File: pkg/services/sqlstore/sqlstore.go [hunk 15]
```diff
@@ -461,6 +461,8 @@ var testSQLStoreMutex sync.Mutex
 type InitTestDBOpt struct {
 	// EnsureDefaultOrgAndUser flags whether to ensure that default org and user exist.
 	EnsureDefaultOrgAndUser bool
+	// Features to turn on.
+	Features []string
```
Comment (ryantxu) [inline pkg/services/sqlstore/sqlstore.go:465]:
```suggestion
	Features featuremgmt.FeatureToggles
```

File: pkg/services/sqlstore/sqlstore.go [hunk 16]
```diff
@@ -500,7 +502,14 @@ func initTestDB(migration registry.DatabaseMigrator, opts ...InitTestDBOpt) (*SQ
...
 		cfg := setting.NewCfg()
-		cfg.IsFeatureToggleEnabled = func(key string) bool { return false }
+		cfg.IsFeatureToggleEnabled = func(key string) bool {
+			for _, f := range opts[0].Features {
+				if f == key {
+					return true
+				}
+			}
+			return false
+		}
```
Comment (ryantxu) [inline pkg/services/sqlstore/sqlstore.go:512]: something like
```suggestion
		cfg.IsFeatureToggleEnabled = opts[0].Features.IsEnabled
```

File: public/app/features/chat/Chat.tsx [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+          ))}
+        </div>
+      );
+    } else {
+      messageElements = <div className="chat-messages">No messages here yet</div>;
+    }
+
+    return (
+      <div
+        className="chat"
```
Comment (nmarrs) [inline public/app/features/chat/Chat.tsx:200]: For custom styles, instead of including in global styles file we can add styling via our `useStyles2` hook. See [docs here](https://github.com/grafana/grafana/blob/main/contribute/style-guides/styling.md). To use hook we would want to convert this to a functional component, which might be worth doing anyways for simplifying structure.

I would recommend following component structure:
`ChatManager` -> top level state management component
`Chat` or something like `ChatContainer` -> What this component currently is but just rendering chat UI / messages, and using passed down state from `ChatManager`
`ChatMessage` - just like you currently are doing, but in a separate file

File: public/app/features/dashboard/components/DashChat/DashChatModal.tsx [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { Modal } from '@grafana/ui';
+import { css } from '@emotion/css';
+import { ChatDashboardModalProps } from './types';
+import { Chat } from 'app/features/chat/Chat';
+
+export const DashboardChatModal: React.FC<ChatDashboardModalProps> = ({ dashboard, onDismiss, onSaveSuccess }) => {
```
Comment (nmarrs) [inline public/app/features/dashboard/components/DashChat/DashChatModal.tsx:7]: No need to include the `onSaveSuccess` if it is not being used

File: public/app/features/dashboard/components/DashChat/DashChatModal.tsx [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+      <Modal
+        isOpen={true}
+        title="Dashboard discussions"
+        icon="save"
+        onDismiss={onDismiss}
+        className={css`
+          width: 500px;
+        `}
+      >
+        {chat}
```
Comment (nmarrs) [inline public/app/features/dashboard/components/DashChat/DashChatModal.tsx:21]: We can have this inline vs. a separate var
```suggestion
        <Chat contentTypeId={2} objectId={dashboard.uid} />
```

File: public/app/features/dashboard/components/DashNav/DashNav.tsx [hunk 20]
```diff
@@ -15,9 +15,11 @@ import { DashboardModel } from '../../state';
 import { KioskMode } from 'app/types';
 import { ShareModal } from 'app/features/dashboard/components/ShareModal';
 import { SaveDashboardModalProxy } from 'app/features/dashboard/components/SaveDashboard/SaveDashboardModalProxy';
+import { DashboardChatModalProxy } from 'app/features/dashboard/components/DashChat/DashChatModalProxy';
```
Comment (nmarrs) [inline public/app/features/dashboard/components/DashNav/DashNav.tsx:18]: Do we need the proxy component in this case? I think we are doing some additional logic for `SaveDashboardModalProxy` that don't apply here

**Comment by nmarrs** [comment]
> But I would not frame this as chat, think chat should be best handled by chat tools like slack etc. I think of this as something very similar/identical to Google doc comments. Think being able to comment & ping someone, and view past comments for a dashboards, alert rules, reports, incidents, etc would be very valuable

Agree with @torkelo on this. Ability to have comments / history of comments similar to google docs could be very valuable (and honestly a revision history of dashboards / panels that you could revert back to specific versions / compare changes as well, but that is a separate feature)

If we do go with `Comment` idea we can replace the naming of `Chat` / `Discussions` with `Comments` (addressing question 1)

**Comment by nmarrs** [comment]
To answer Q4
> Can we re-initialize widget on connection re-establishement?

Yes we can, as we can include logic as part of a `useEffect` in the `CommentManagement` component that listens to a subscription and on connection change re-renders comments accordingly

File: public/app/features/chat/Chat.tsx [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+      </MenuGroup>
+    );
+  };
+  return (
+    <div
+      onMouseEnter={() => {
+        setShowActionIcon(true);
+      }}
+      onMouseLeave={() => {
+        setShowActionIcon(false);
```
Comment (nmarrs) [inline public/app/features/chat/Chat.tsx:268]: A bit cleaner to pull these inline event handlers into separate defined functions above
e.g.
```
const onMouseEnter = () => {
    setShowActionIcon(false);
}
...

  onMouseEnter={onMouseEnter}
```

File: public/app/features/chat/Chat.tsx [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+      }}
+      className="chat-message"
+    >
+      <div style={{ float: 'left', paddingTop: '6px', marginRight: '10px' }}>
+        <img src={avatarUrl} alt="" style={{ width: '30px', height: '30px' }} />
+      </div>
+      <div style={{ position: 'relative' }}>
```
Comment (nmarrs) [inline public/app/features/chat/Chat.tsx:275]: Can use [docs here](https://github.com/grafana/grafana/blob/main/contribute/style-guides/styling.md) and create a `getStyles` func that returns classes vs. doing inline styling here

Inline styling for colors is fine given they are dynamic

File: public/app/features/chat/Chat.tsx [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+      scope: LiveChannelScope.Grafana,
+      namespace: 'chat', // holds on to the last value
+      path: `${this.props.contentTypeId}/${this.props.objectId}`,
+    };
+  };
+
+  getLiveChannel = () => {
+    const live = getGrafanaLiveSrv();
+    if (!live) {
+      console.error('Grafana live not running, enable "live" feature toggle');
```
Comment (nmarrs) [inline public/app/features/chat/Chat.tsx:109]: In general we will want to go through and add some error handling UX on the event the service is down / user is not connected for whatever reason and whatever other edge cases we can think of.

Given this is early not necessarily a blocker for this PR, but would be good to consider for when we polish this to go into beta / beyond

**Comment by nmarrs** [comment]
Would it be possible to update PR description and include a few screenshots / screencast (or maybe even a link to the ChaChaChat hackathon submission vid? (which I believe this originated?) For anything involving frontend it's always nice to have some form of visual for context

File: public/app/features/chat/Chat.tsx (hunk 9 above)
Comment (FZambia) [reply public/app/features/chat/Chat.tsx:42]: Existing funcs use the following markdown options:

```
const markdownOptions = {
  pedantic: false,
  gfm: true,
  smartLists: true,
  smartypants: false,
  xhtml: false,
};
```

In live discussions we use the following currently:

```
{
    pedantic: false,
    gfm: true,
    smartLists: true,
    smartypants: false,
    xhtml: false,
    breaks: true,
}
```

I.e. adding `breaks: true` to correctly render line breaks.

File: pkg/services/sqlstore/sqlstore.go (hunk 16 above)
Comment (FZambia) [reply pkg/services/sqlstore/sqlstore.go]: Done

File: pkg/services/sqlstore/sqlstore.go (hunk 15 above)
Comment (FZambia) [reply pkg/services/sqlstore/sqlstore.go]: Done

File: pkg/api/api.go (hunk 12 above)
Comment (FZambia) [reply pkg/api/api.go]: Removed

File: pkg/services/featuremgmt/registry.go (hunk 13 above)
Comment (FZambia) [reply pkg/services/featuremgmt/registry.go]: Removed

File: pkg/api/index.go (hunk 3 above)
Comment (FZambia) [reply pkg/api/index.go]: removed

File: public/app/features/chat/routes.tsx (hunk 8 above)
Comment (FZambia) [reply public/app/features/chat/routes.tsx]: removed

File: pkg/services/featuremgmt/registry.go (hunk 14 above)
Comment (FZambia) [reply pkg/services/featuremgmt/registry.go]: Done

File: public/app/features/chat/Chat.tsx [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+  verbal: string;
+  action: any;
+}
+
+interface ChatMessageProps {
+  message: Message;
+  actions?: ChatMessageAction[];
+}
+
+const ChatMessage: FunctionComponent<ChatMessageProps> = ({ message, actions = [] }) => {
```
[the remark this answers is not in the document]
Comment (FZambia) [reply public/app/features/chat/Chat.tsx]: Done

File: public/app/features/dashboard/components/DashChat/DashChatModal.tsx (hunk 19 above)
Comment (FZambia) [reply public/app/features/dashboard/components/DashChat/DashChatModal.tsx]: Done

File: public/app/features/dashboard/components/DashChat/DashChatModal.tsx (hunk 18 above)
Comment (FZambia) [reply public/app/features/dashboard/components/DashChat/DashChatModal.tsx]: Done

File: public/app/features/chat/Chat.tsx (hunk 21 above)
Comment (FZambia) [reply public/app/features/chat/Chat.tsx]: Done

File: public/app/features/dashboard/components/DashNav/DashNav.tsx (hunk 20 above)
Comment (FZambia) [reply public/app/features/dashboard/components/DashNav/DashNav.tsx]: Works without it, removed

File: public/app/features/chat/Chat.tsx (hunk 17 above)
Comment (FZambia) [reply public/app/features/chat/Chat.tsx:172]: In this case, to control scroll and input focus of `ChatContainer ` from `ChatManager` we should use child refs, like:

```
this.refs.chatContainer.refs.chatInput.focus();
```

?

Or there is a better way?

**Comment by torkelo** [comment]
Not sure the naming is worse with the word `comment`, feels more like an entity name (suitable table name) compared to chat 

Example naming with comment
* Table: comment  (looks like our ables are singular)
* Entity: Comment or CommentItem 
  
UI telements
* CommentsBox or EntityComments , or DocComments 
* CommentItem  (single message)

**Comment by ArturWierzbicki** [comment]
> > Not sure the naming is worse with the word comment, feels more like an entity name (suitable table name) compared to chat
> 
> @torkelo we have 2 new database entities/tables at the moment (`Chat` and `ChatMessage`, chat has many chat messages), my confusion is about how to name both in this case.

It looks like we need a table which stores metadata related to a group of messages - maybe lets go with `message` and `message_group` (or replace `message` with `comment`)?

**Comment by torkelo** [comment]
Maybe CommentThread / comment_thread for settings relating to a group of comments

**Comment by torkelo** [comment]
But maybe implies nesting?

**Comment by torkelo** [comment]
> It looks like we need a table which stores metadata related to a group of messages - maybe lets go with message and message_group (or replace message with comment)?

comment & comment_group could work

File: packages/grafana-data/src/types/featureToggles.gen.ts [hunk 25]
```diff
@@ -39,4 +39,6 @@ export interface FeatureToggles {
   validatedQueries?: boolean;
   swaggerUi?: boolean;
   featureHighlights?: boolean;
+  liveDashboardDiscussions?: boolean;
```
Comment (torkelo) [inline packages/grafana-data/src/types/featureToggles.gen.ts:42]: not sure this needs to be related to live specifically? That seems more like an implementation detail. The feature is just `comments`

File: public/app/features/chat/Chat.tsx [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+      this.setState(
+        (prevState) => ({ value: prevState.value.substring(0, start) + '\n' + prevState.value.substring(end) }),
+        () => {
+          this.chatInput.selectionStart = this.chatInput.selectionEnd = start + 1;
+        }
+      );
+    } else if (e.key === 'Enter') {
+      e.preventDefault();
+    }
+  };
```
Comment (torkelo) [inline public/app/features/chat/Chat.tsx:106]: Not sure what this is does with the selection? 

Nitpick: do not think enter key should submit the message. Enter is useful for newlines and formatting the message. A button is likely better.

File: public/app/features/chat/Chat.tsx (hunk 17 above)
Comment (torkelo) [reply public/app/features/chat/Chat.tsx:172]: when do you want to focus input? looks like its only on mount, could it not be handled in the component? If you need to control it from the parent you could do it by passing a property "focusCounter" when it changes the component will re-focus the input

File: public/app/features/chat/ChatMessage.tsx [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+  });
+  return sanitize(html);
+}
+
+interface ChatMessageProps {
+  message: Message;
+  actions?: ChatMessageAction[];
+}
+
+export const ChatMessage: FunctionComponent<ChatMessageProps> = ({ message, actions = [] }) => {
```
Comment (torkelo) [inline public/app/features/chat/ChatMessage.tsx:33]: You do not need to use `: FunctionComponent<ChatMessageProps> ` typing. It's not considered good practice anymore. Just write 

ts
```
export function ChatMessage({ message, actions = [] }: Props) { 
}
```

instead and rely on implicit typing

File: public/app/features/chat/ChatMessage.tsx [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+  const html = marked(str || '', {
+    pedantic: false,
+    gfm: true,
+    smartLists: true,
+    smartypants: false,
+    xhtml: false,
+    breaks: true,
+  });
+  return sanitize(html);
+}
```
Comment (torkelo) [inline public/app/features/chat/ChatMessage.tsx:26]: use renderMarkdown from `@grafana/data`

File: packages/grafana-data/src/text/markdown.ts [hunk 29]
```diff
@@ -21,7 +23,15 @@ export function renderMarkdown(str?: string, options?: RenderMarkdownOptions): s
     hasInitialized = true;
   }
 
-  const html = marked(str || '');
+  let opts = undefined;
+  if (options?.breaks) {
+    opts = {
+      ...markdownOptions,
+      breaks: true,
```
Comment (ryantxu) [inline packages/grafana-data/src/text/markdown.ts:30]: ideally we could just set `breaks: true` for everything -- but that may cause issues with existing text panels!

File: pkg/services/chats/chatmodel/models.go [hunk 30]
```diff
@@ -0,0 +1,71 @@
+package chatmodel
+
+const (
+	// ContentTypeOrg is reserved for future use for per-org discussions.
+	ContentTypeOrg = 1
+	// ContentTypeDashboard used for dashboard-wide discussions.
+	ContentTypeDashboard = 2
```
Comment (ryantxu) [inline pkg/services/chats/chatmodel/models.go:7]: the magic numbers are a bit confusing and maybe limited... maybe better would be constant strings.  For example:
* dash
* anno

comments on individual panels seems valuable -- but also nervous about a page with a subscription for each panel.

File: public/app/features/chat/Chat.tsx (hunk 26 above)
Comment (nmarrs) [reply public/app/features/chat/Chat.tsx]: Just updated / cleaned up frontend, did not include button at this time, however user can make newlines via "shift" + "enter".

On a whole I think that "enter" to submit the comment is inline with UX paradigms (similar to Slack's UX). As we refine this I think that we may want to include a separate send button eventually (for accessibility sake alone) as well as ability to edit / delete comments etc.

**Comment by nmarrs** [comment]
Went through (pairing with @ryantxu) and cleaned up the frontend. Did another pass through of frontend to clean up a few missed loose ends.

From my perspective, this is getting very close to mergable. Pretty much the last thing to address in my opinion is updating the backend naming scheme to use "comments" vs. chat / discussions (already done in frontend except for connection into backend data in `CommentManager` component)

**Comment by FZambia** [comment]
Renamed backend parts and API layer to use `comments`.

File: packages/grafana-data/src/types/featureToggles.gen.ts (hunk 25 above)
Comment (FZambia) [reply packages/grafana-data/src/types/featureToggles.gen.ts]: Fixed, + updated PR description

File: pkg/services/chats/chatmodel/models.go (hunk 30 above)
Comment (FZambia) [reply pkg/services/chats/chatmodel/models.go]: Using strings for content types now, in database table too: NVarChar(10). And strings in API layer.

File: pkg/services/comments/commentmodel/models.go [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+var RegisteredContentTypes = map[string]struct{}{
+	ContentTypeOrg:        {},
+	ContentTypeDashboard:  {},
+	ContentTypeAnnotation: {},
+}
+
+type CommentGroup struct {
+	Id          int64
+	OrgId       int64
+	ContentType string
```
[the remark this answers is not in the document]
Comment (FZambia) [reply pkg/services/comments/commentmodel/models.go:28]: That's rather common in Django world: https://docs.djangoproject.com/en/4.0/ref/contrib/contenttypes/. I think if we want to change then maybe `ObjectType` or `EntityType` are better.

**Comment by nmarrs** [comment]
> I noticed 2 problems with the latest changes:
> 
> 1. Comment write textarea outline not displayed correctly.
> 2. Scrolling to the bottom does not work for annotations

@FZambia latest commit addresses 2 issues above, did another review of frontend and things are looking good to me for this initial version 👌

File: pkg/services/comments/commentmodel/models.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (FZambia) [reply pkg/services/comments/commentmodel/models.go]: Renamed to `ObjectType` in dc9c66a14517a3d3b9070b2a81052fcd77876821 since we already have `ObjectId` nearby. Is it ok?

**Comment by FZambia** [comment]
@torkelo hello, can you please suggest, what's the best place for dashboard comments icon? It's currently in between `save` and `settings` (as shown on video in pr description) – I suppose we better move it?

**Comment by torkelo** [comment]
@FZambia think it makes sense to have it next t the share button on the right side 
Screen Shot 2022-02-16 at 14 26 21

**Comment by torkelo** [comment]
Can we use the icon [comment-alt-message ](https://iconscout.com/unicons/explore/line#) think that works best

**Comment by torkelo** [comment]
Screen Shot 2022-02-16 at 14 29 26

**Comment by FZambia** [comment]
@torkelo , looks like this:

Screenshot 2022-02-16 at 17 01 34

Had to add `comment-alt-message` to `grafana-ui` supported IconName values - hope it's ok.

**Comment by torkelo** [comment]
Looks good!

**Comment by sdboyer** [comment]
we're trying to get rid of the bus in #44130, and have already made a service available for dashboards in #44923. Could we follow those patterns here, rather than introducing a new bus dependency we'll need to shortly remove?

**Review by ryantxu — APPROVED** [review]
@FZambia -- this is a good place to let us start testing it for real.  Can you make the changes to bus suggested by @sdboyer and merge next week?

File: pkg/services/comments/handlers.go [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+
+	userMap := map[int64]*models.UserSearchHitDTO{}
+	if signedInUser.UserId > 0 {
+		q := &models.SearchUsersQuery{Query: "", Filters: []models.Filter{NewIDFilter([]int64{signedInUser.UserId})}, Page: 0, Limit: 1}
+		if err := s.sqlStore.SearchUsers(ctx, q); err != nil {
+			return nil, err
+		}
+		for _, u := range q.Result.Users {
+			userMap[u.Id] = u
+		}
```
Comment (torkelo) [inline pkg/services/comments/handlers.go:100]: why are you searching for the signed-in user?

Comment (torkelo) [inline pkg/services/comments/handlers.go:100]: why are you searching for the signed-in user?

File: pkg/services/comments/commentmodel/models.go [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+	return "comment_group"
+}
+
+type Settings struct {
+}
+
+var (
+	_ driver.Valuer = Settings{}
+	_ sql.Scanner   = &Settings{}
+)
```
Comment (torkelo) [inline pkg/services/comments/commentmodel/models.go:46]: have not seen this anywhere else, can you describe what this is?

Comment (FZambia) [reply pkg/services/comments/commentmodel/models.go:46]: This is a way to work with JSON using a typed approach - i.e. we can use structs in Go, but it will be automatically marshalled to JSON when saving into the database and unmarshalled when retrieving.

**Review by torkelo — APPROVED** [review]

File: pkg/services/comments/handlers.go (hunk 32 above)
Comment (torkelo) [reply pkg/services/comments/handlers.go:100]: @FZambia we get avatarUrl in index.go from just SignInUser 

dtos.GetGravatarUrl(c.Email) 

https://github.com/grafana/grafana/blob/main/pkg/api/index.go#L30

Comment (FZambia) [reply pkg/services/comments/handlers.go:100]: I see - will try to avoid this then.

Comment (FZambia) [reply pkg/services/comments/handlers.go]: done in 6cfb62773187466371a65ceb5c76f756110e6352
